File: spark/spark-examples/src/main/scala/com/aliyun/spark/polardb/SparkOnPOLARDBSparkSession.scala
package com.aliyun.spark.polardb

import java.util.Properties

import org.apache.spark.sql.SparkSession

object SparkOnPOLARDBSparkSession {

  def main(args: Array[String]): Unit = {
    //获取POLARDB的 url、database、tableName、登录POLARDB数据库的user和password
    val url = args(0)
    val jdbcConnURL = s"jdbc:mysql://$url"
    val database = args(1)
    val tableName = args(2)
    val user = args(3)
    val password = args(4)

    //Spark侧的表名。
    var sparkTableName = if (args.size > 5) args(5) else "spark_on_polardb_sparksession_test01"

    val sparkSession = SparkSession
      .builder()
      .enableHiveSupport() //可选，使用hive-metastore后通过thriftServer可以查看到代码中创建的表
      .appName("scala spark on POLARDB test")
      .getOrCreate()

    //Sql方式，Spark会映射POLARDB中表的Schema。
    sparkTableName = sparkTableName + "_noschema"
    var createCmd =
      s"""CREATE TABLE ${sparkTableName} USING org.apache.spark.sql.jdbc
         |    options (
         |    url '$jdbcConnURL',
         |    dbtable '$database.$tableName',
         |    user '$user',
         |    password '$password'
         |    )""".stripMargin

    sparkSession.sql(createCmd)
    var querySql = "select * from " + sparkTableName + " limit 1"
    sparkSession.sql(querySql).show


    //使用dataset API接口
    val connectionProperties = new Properties()
    connectionProperties.put("user", user)
    connectionProperties.put("password", password)
    //读取数据
    var jdbcDf = sparkSession.read.jdbc(jdbcConnURL, s"$database.$tableName", connectionProperties)
    jdbcDf.select("name", "age", "score").show()

    val data =
      Seq(
        Person("bill", 30, 170.5),
        Person("gate", 29, 200.3)
      )
    val dfWrite = sparkSession.createDataFrame(data)

    //写入数据
    dfWrite
      .write
      .mode("append")
      .jdbc(jdbcConnURL, s"$database.$tableName", connectionProperties)
    jdbcDf.select("name", "age").show()
    sparkSession.stop()
  }

}

case class Person(name: String, age: Int, score: Double)

File: spark/python/polardb/SparkOnPolardbSparkSession.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
"""
X-Pack Spark 控制台提交命令示例

--driver-memory 1G
--driver-cores 1
--executor-cores 1
--executor-memory 1G
--num-executors 1
--jars /mysql-connector-java-5.1.34.jar
--name on_polardb
/SparkOnPolardbSparkSession.py
${polardb的访问地址} ${polardb的数据库名} ${polardb的表名} ${登录polardb数据库的用户名} ${登录polardb数据库的密码} ${Spark侧的表名}
"""
import sys
from pyspark.sql import SparkSession

if __name__ == "__main__":

    #  获取POLARDB的 url、database、tableName、登录POLARDB数据库的user和password
    url = sys.argv[1]
    jdbcConnURL = "jdbc:mysql://" + url
    database = sys.argv[2]
    tableName = sys.argv[3]
    user = sys.argv[4]
    password = sys.argv[5]
    driver = "com.mysql.jdbc.Driver"
    # Spark侧的表名。
    sparkTableName = sys.argv[6]

    spark = SparkSession\
        .builder\
        .appName("spark on hbase")\
        .getOrCreate()

    create_cmd = ('CREATE TABLE ' + sparkTableName + ' USING org.apache.spark.sql.jdbc\n'
             '	OPTIONS (\n'
             '      driver \'' + driver + '\',\n'
             '      url \'' + jdbcConnURL + '\',\n'
             '      dbtable \'' + database + '.' + tableName + '\',\n'
             '      user \'' + user + '\',\n'
             '      password \'' + password + '\'\n'
             '	)')
    print(create_cmd)
    spark.sql(create_cmd)
    query = "select * from " + sparkTableName + " limit 10"
    print(query)
    spark.sql(query).show()
    spark.stop()